### ISSUE TITLE
Temporal Dead Zone variable usage, new-rule/rule-change

### ISSUE THREAD

**Issue by tchetwin** [eslint/eslint#11430]
I was a little surprised to find that usage of a variable within its [temporal dead zone](http://2ality.com/2015/10/why-tdz.html) (henceforth "TDZ") is not considered to be an error in the `eslint:recommended` ruleset.

Example of TDZ:

```javascript
{            // <-- block scope, not function scope
    foo;     // <--+
             //    |-- TDZ for `foo`
             // <--+
    let foo; // <-- ReferenceError: foo is not defined
}
```

Looking back in the archives I see a couple of relevant closed Issues:

* #10375: concluded that `no-use-before-define` is sufficient
* #9677: timed out

I think this space would benefit from either a new rule, or some adjustment. The rest of this post outlines the existing situation and possible solutions. Feedback welcomed!

---

## Existing rule, `no-use-before-define`

This rule is not enabled by the ruleset `eslint:recommended`.

The rule can catch TDZ usage, see [this ESLint demo](https://eslint.org/demo/#eyJ0ZXh0IjoieyAgICAgICAgICAgIC8vIDwtLSBibG9jayBzY29wZSwgbm90IGZ1bmN0aW9uIHNjb3BlXHJcbiAgICBmb287ICAgICAvLyA8LS0rXHJcbiAgICAgICAgICAgICAvLyAgICB8LS0gVERaIGZvciBgZm9vYFxyXG4gICAgICAgICAgICAgLy8gPC0tK1xyXG4gICAgbGV0IGZvbzsgLy8gPC0tIFJlZmVyZW5jZUVycm9yOiBmb28gaXMgbm90IGRlZmluZWRcclxufSIsIm9wdGlvbnMiOnsicGFyc2VyT3B0aW9ucyI6eyJlY21hVmVyc2lvbiI6Niwic291cmNlVHlwZSI6InNjcmlwdCIsImVjbWFGZWF0dXJlcyI6e319LCJydWxlcyI6eyJuby11c2UtYmVmb3JlLWRlZmluZSI6Mn0sImVudiI6e319fQ==).

However, it [also catches](https://eslint.org/demo/#eyJ0ZXh0IjoieyAgICAgICAgICAgIC8vIDwtLSBibG9jayBzY29wZSwgbm90IGZ1bmN0aW9uIHNjb3BlXHJcbiAgICBmb287ICAgICAvLyA8LS0rXHJcbiAgICAgICAgICAgICAvLyAgICB8LS0gYGZvb2AgZGVjbGFyYXRpb24gaXMgaG9pc3RlZCwgdmFsdWUgaXMgYHVuZGVmaW5lZGBcclxuICAgICAgICAgICAgIC8vIDwtLStcclxuICAgIHZhciBmb287XHJcbn0iLCJvcHRpb25zIjp7InBhcnNlck9wdGlvbnMiOnsiZWNtYVZlcnNpb24iOjYsInNvdXJjZVR5cGUiOiJzY3JpcHQiLCJlY21hRmVhdHVyZXMiOnt9fSwicnVsZXMiOnsibm8tdXNlLWJlZm9yZS1kZWZpbmUiOjJ9LCJlbnYiOnt9fX0=) similar `var` usage, at which point we have ventured into subjective territory.

It is possible that people wish to use `var` hoisting in this manner. It is *very unlikely* that people want to use TDZ `ReferenceError` as part of control flow.

### (1) New Rule, e.g. `no-temporal-dead-zone-usage`

A new rule could be created, and I hope included in the `eslint:recommended` ruleset to benefit the masses.

It would notify users of situations where a `let`/`const` variable is referenced before its declaration.

### (2) Extend existing `no-use-before-define`

Parameters could be added to the existing rule to support this use case. I haven't fleshed out what these might look like, possibly an additional property. Would have to consider how folks might enable the TDZ checking without also enabling `var` hoisting errors.

I would hope that the TDZ checking could be made part of `eslint:recommended` via this avenue too, though this may be trickier.

### Comparison

With respect to the [Core Rule Guidelines](https://eslint.org/docs/developer-guide/contributing/new-rules#core-rule-guidelines), both approaches are:

* Widely applicable ✔️ 
* Generic ✔️ 
* Atomic ✔️ 
* Library agnostic ✔️ 

An additional rule might fall foul of Uniqueness and "No conflicts".

## Issue Template

**What category of rule is this?**

[X] Warns about a potential error (problem)
[ ] Suggests an alternate way of doing something (suggestion)
[ ] Enforces code style (layout)
[ ] Other (please specify:)

**Why should this rule be included in ESLint (instead of a plugin)?**

This rule would warn about JavaScript language use and would ideally be included in the `eslint:recommended` ruleset. It would detect usage that is almost universally a bug.

**Are you willing to submit a pull request to implement this rule?**

Yes

**Does this change cause the rule to produce more or fewer warnings?**

Depends on configuration. It would cause fewer warnings for someone wishing to block TDZ usage but wanting to use `var` hoisting.

**Comment by tchetwin** [comment]
**Edit:** Somehow one of my links above to ESLint demos was incorrect. I have updated the post.

In case it's still unclear the two examples were:

```javascript
{            // <-- block scope, not function scope
    foo;     // <--+
             //    |-- TDZ for `foo`
             // <--+
    let foo; // <-- ReferenceError: foo is not defined
}
```

```javascript
{            // <-- block scope, not function scope
    foo;     // <--+
             //    |-- `foo` declaration is hoisted, value is `undefined`
             // <--+
    var foo;
}
```

With only the rule `no-use-before-define` turned on. Both result in an error. The former error is wholly justified. The latter is subjective.

**Comment by mysticatea** [comment]
Thank you for this proposal.

I think that `no-undef` rule is the best rule to report the clear TDZ accesses.

However, we cannot report all TDZ errors in the static analysis because TDZ is a timing matter.

```js
{
    foo(); // TDZ ReferenceError: `x` is not defined.
    [0].forEach(foo); // TDZ ReferenceError: `x` is not defined.
    setTimeout(foo, 0); // OK!
    let x = 0;
    foo(); // OK!

    function foo() { x } //← access `x`
}
```

**Comment by tchetwin** [comment]
Great catch, I had not considered the *temporal* aspect sufficiently, thanks for taking the time to reply to my Issue.

As you point out it's probably beyond the scope of syntactic analysis to find all the TDZ cases. If I'm not mistaken there are still cases we can statically analyse without false positives, and it would be good not to [throw the baby out with the bath water](https://en.wikipedia.org/wiki/Don%27t_throw_the_baby_out_with_the_bathwater):

```javascript
function foo () {
    console.log(x);
}

x;     // ReferenceError, analysable

foo(); // ReferenceError, beyond our grasp!

let x;
```

I believe the [`no-undef`](https://eslint.org/docs/rules/no-undef) rule does not currently cater to this use case, are you suggesting that it would be a more appropriate rule to extend?

**Comment by mysticatea** [comment]
> I believe the [`no-undef`](https://eslint.org/docs/rules/no-undef) rule does not currently cater to this use case, are you suggesting that it would be a more appropriate rule to extend?

Yes. There are some reasons.

- `no-use-before-define` rule has been separated from `no-undef` rule because `var` declarations don't throw `ReferenceError` even if we accessed the variable before `var` declarations. (I.e., it's a different behavior to the access to undefined variables.) (This happened before 2015.)
- On the other hand, the variables which are defined by any of `let`, `const`, and `class` throw `ReferenceError` as same as the access to undefined variables if we accessed to those in TDZ.
- `eslint:recommended` contains `no-undef` rule.
